### PR TITLE
make Constrain Hits first class citizen

### PR DESCRIPTION
**Pull request by VinInn**
This PR makes Transient Constrain Hits "first class citizen" so that they can be used in standard fits and pattern recognition.
Possible applications of the 5D Constrain can be to represent a "tracklet" (a seed for instance) to be used "asis" during CKF (think of L1 trackless or pixelTrack from FPGA or GPU) . 
2D Constrains can be used as "Frozen Hits" w/o any further reference to clusters.

As the required mods touch very basic packages, here is a purely technical PR w/o ay further infrastructure to try/test any of the use case above.
This will allow further development w/o the need of recompiling half of the universe.

purely technical NO regression expected.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
doubt this problem is caused by this PR
```
===== Test "testSSTGainPCL_fromRECO" ====
argument 0: testSSTGainPCL_fromRECO
argument 1: /bin/bash
argument 2: CalibTracker/SiStripChannelGain/test
argument 3: testSSTGain_PCL_FromRECO.sh
shell is: /bin/bash
Current directory is: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100
topdir is: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100
testdir is: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100/src/CalibTracker/SiStripChannelGain/test
tmpdir is: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100/tmp/slc6_amd64_gcc700
testbin is: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100/test/slc6_amd64_gcc700
Running script: /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100/src/CalibTracker/SiStripChannelGain/test/testSSTGain_PCL_FromRECO.sh
Using Dataset: /ZeroBias/Run2018A-SiStripCalMinBias-17Sep2018_pilot-v1/ALCARECO
----- Begin Fatal Exception 24-Nov-2018 20:12:18 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_4_X_2018-11-24-1100/src/CalibTracker/SiStripChannelGain/test/testSSTGain_PCL_FromRECO_cfg.py
Exception Message:
python encountered the error: <type 'exceptions.ValueError'>
No JSON object could be decoded
----- End Fatal Exception -------------------------------------------------
Failure running cmsRun testSSTGain_PCL_FromRECO_cfg.py 0: status 90
status = 23040

---> test testSSTGainPCL_fromRECO had ERRORS
 
^^^^ End Test testSSTGainPCL_fromRECO ^^^^
```

**Comment by fabiocos** [comment]
@VinInn this problem has appeared in recent IBs but I was unable to reproduce it offline

**Comment by fabiocos** [comment]
@VinInn ok, I reproduce it (I must have done something wrong the first time)

**Comment by mmusich** [comment]
given the last IB is clean:
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc6_amd64_gcc700/CMSSW_10_4_X_2018-11-27-1100/unitTestLogs/CalibTracker/SiStripChannelGain#/
I guess tests can be restarted here

**Comment by andrius-k** [comment]
+1

**Comment by slava77** [comment]
@VinInn 
how urgent is this PR?
The changes in ```trackerHitRTTI``` here conflict with the changes in #25063  (and to make matters worse MTD has opened also  #25384 with the same changes).

Specifically, ```inline bool isFromDet(TrackingRecHit const & hit)  { return (rtti(hit)>0) & (rtti(hit)<6) ;}``` Should it eventually include the MTD hits or not?


If this PR can go on 10_5_X trajectory, I propose to wait until then.

File: RecoTracker/TrackProducer/interface/TrackProducerBase.icc [hunk 1]
```diff
@@ -180,8 +180,9 @@ TrackProducerBase<T>::setSecondHitPattern(Trajectory* traj, T& track,
       //means  that the trajectory was fit/smoothed in a special case: not setting those pointers
 	edm::LogError("TrackProducer") << "the trajectory was fit/smoothed in a special case: not setting those pointers.\n"
 			<<" Filling the secondary hit patterns was requested. So I will bail out.";
-	throw cms::Exception("TrackProducerBase")<<"the trajectory was fit/smoothed in a special case: not setting those pointers.\n"
-			<<" Filling the secondary hit patterns was requested. So I will bail out.";
+//	throw cms::Exception("TrackProducerBase")<<"the trajectory was fit/smoothed in a special case: not setting those pointers.\n"
+//			<<" Filling the secondary hit patterns was requested. So I will bail out.";
```
[the remark this answers is not in the document]
Comment (VinInn) [reply RecoTracker/TrackProducer/interface/TrackProducerBase.icc]: I am not sure that an exception is needed here, after all we are in a core tracking code...
anyhow it works now with the new RTTI so we can keep the exception in....

File: DataFormats/TrackerRecHit2D/interface/trackerHitRTTI.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  inline bool isSingle(TrackingRecHit const & hit)  { return rtti(hit)==single || rtti(hit)==fastSingle;}
+  inline bool isProjMono(TrackingRecHit const & hit)  { return rtti(hit)==projMono || rtti(hit)==fastProjMono;}
+  inline bool isProjStereo(TrackingRecHit const & hit)  { return rtti(hit)==projStereo || rtti(hit)==fastProjStereo;}
+  inline bool isProjected(TrackingRecHit const & hit)  { return ((rtti(hit)==projMono) | (rtti(hit)==projStereo)) || (rtti(hit)==fastProjMono) | (rtti(hit)==fastProjStereo);}
+  inline bool isMatched(TrackingRecHit const & hit)  { return rtti(hit)==match || rtti(hit)==fastMatch;}
+  inline bool isMulti(TrackingRecHit const & hit)  { return rtti(hit)==multi;}
+  inline bool isSingleType(TrackingRecHit const & hit)  { return (rtti(hit)>0) & (rtti(hit)<4) ;}
+  inline bool isFromDet(TrackingRecHit const & hit)  { return (rtti(hit)>0) & (rtti(hit)<6) ;}
+  inline bool isFast(TrackingRecHit const & hit)  { return (rtti(hit)>5) & (rtti(hit)<=9) ;}
+  inline bool isFromDetOrFast(TrackingRecHit const & hit)  { return (rtti(hit)>0) & (rtti(hit)<10) ;}
```
[the remark this answers is not in the document]
Comment (VinInn) [reply DataFormats/TrackerRecHit2D/interface/trackerHitRTTI.h]: indeed. there is space, but if we renumber we break compatibility with previous Reco.
Btw this is something to discuss form Run3:
1) can we assume that the "Reco" tier is no more persistent. only AOD is persistent and we can have much more freedom with whatever is not saved in AOD?
2) in any case can we break backward compatibility for anything that is not saved in AOD?

**Comment by VinInn** [comment]
I would like to have in for Patatrack development and more in general HLT (or fast reco) for Run3 and Phase2.
The idea is to have a seed (or whatever) already fitted and continue from there.
I tested, with this infrastructure goes till MTV...

**Comment by VinInn** [comment]
for what concern "mipTiming=10" I suspect the second who goes in will have to renumber....

**Comment by slava77** [comment]
On 12/2/18 8:09 AM, Vincenzo Innocente wrote:
> Btw this is something to discuss form Run3:
> 
>  1. can we assume that the "Reco" tier is no more persistent. only AOD
>     is persistent and we can have much more freedom with whatever is not
>     saved in AOD?


>  2. in any case can we break backward compatibility for anything that is
>     not saved in AOD?

RAW-RECO skims are still made routinely.
This should be considered in the decision process to try to get more
freedom in the data transient with respect to AOD.

As usual, some cost-benefit analysis is necessary.
Are these questions in the context of this PR or in general?
For "in general", this may be not the right place to discuss.

If a compatibility has to be broken, perhaps this should be done before
the ultra-legacy release is made.

**Comment by slava77** [comment]
On 12/2/18 8:15 AM, Vincenzo Innocente wrote:
> for what concern "mipTiming=10" I suspect the second who goes in will
> have to renumber....
> 
> —

It looks to me that it's not as trivial as just renumbering.
In a few places there are decisions made to skip the hit or process it.
Is an MTD hit "fromDet" or not?

**Comment by VinInn** [comment]
> On 2 Dec, 2018, at 8:04 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> It looks to me that it's not as trivial as just renumbering.
> In a few places there are decisions made to skip the hit or process it.
> Is an MTD hit "fromDet" or not?
No clue, its use need to be reviewed... 
I already added a messy fromDetOrFast not to enter in discussions with FastSim.
In any case all this special cases will not scale, DPG,DQM and TRK (MUON!!!) need to review what hit shall be considered for each use case
and eventually come with a new classification
(isGoodForFit, isGoodForRefit, isGoodForPatternReco, isGoodForSeeding, haveOmniRef,  etc)
and need to be fast, no C++ RTTI, and at some point encapsulated avoiding to pollute all possible tracking packages

In reality for Phase2 the whole hit infrastructure may change, even if hits sub-det will have its pecularity
InnerTracker: HIts frozen for Seeds, refit for precision fitting
OuterTracker: Just frozen, no ref to cluster?
MTD: always refit (info limited, maybe no need of a Ref, just cache raw info)
(in any case MTD will need some optimization of data structure, now looks more for a test-beam or sim-studies)

All this can be done very differenlty than the current DataFormat,
still I suspect we will need to support this structure for the next 5-7 years...

NOW: what is the priority? Run3, TDRs or Phase2?
and how to coordinate (MTD DPG is clearly touching TRK-POG software w/o any consultation, fine for R&D, not so fine for code that will have to coesist wirh Run2-Legacy and Run3 development)

v.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
assign upgrade


there appear to be some implications on phase-2 integration with the changes added here

**Comment by andrius-k** [comment]
please test

**Comment by kpedro88** [comment]
@lgray @pmeridian please be aware of this PR

**Comment by slava77** [comment]
There are issues with headers:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25333/31947/headers_chks.log
```RecoTracker/TransientTrackingRecHit/interface/TRecHit5DParamConstraint.h:63:11: error: 'TransientTrackingRecHit' does not name a type```

Please check and resolve

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
the last update did not resolve all errors https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25333/32004/headers_chks.log

```
RecoTracker/TransientTrackingRecHit/interface/TRecHit5DParamConstraint.h:73:34: 
error: field 'tsos_' has incomplete type 'const TrajectoryStateOnSurface'
   const TrajectoryStateOnSurface tsos_;
```

It looks like the commands at the top of the log file can be used to check the problems locally.

**Comment by VinInn** [comment]
why this header stuff is not run in the code-checks?

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
> why this header stuff is not run in the code-checks?

@smuzaffar 
should we at add it to the cms-bot summary message on github?
Current path of detection is to notice it in the summary.html in the test results

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/25333	15e5f1b
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences as expected.

It looks like a possible conflict with the MTD changes can be easily resolved.

**Comment by civanch** [comment]
+1

**Comment by andrius-k** [comment]
+1

**Comment by slava77** [comment]
@kpedro88 
please check and sign this if it's OK to go in now (I added the upgrade signature due to an apparent overlap with MTD).

**Comment by kpedro88** [comment]
+upgrade

**Comment by slava77** [comment]
@fabiocos 
please clarify why this is still not merged

**Comment by fabiocos** [comment]
+1

@slava77 @kpedro88 @lgray this will trigger a conflict in #25063, but as far as I can see it should be simple to adjust

**Pull request merged**